upgrade-agent/server.go

```diff
@@ -0,0 +1,151 @@
+/*
+Copyright (c) Edgeless Systems GmbH
+
+SPDX-License-Identifier: AGPL-3.0-only
+*/
+
+package upgradeagent
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"os"
+	"os/exec"
+	"regexp"
+
+	"github.com/edgelesssys/constellation/v2/internal/constants"
+	"github.com/edgelesssys/constellation/v2/internal/file"
+	"github.com/edgelesssys/constellation/v2/internal/installer"
+	"github.com/edgelesssys/constellation/v2/internal/logger"
+	"github.com/edgelesssys/constellation/v2/internal/versions"
+	"github.com/edgelesssys/constellation/v2/upgrade-agent/upgradeproto"
+	"google.golang.org/grpc"
+)
+
+var versionRegexp = regexp.MustCompile(`^v\d{1}\.\d{1,2}\.\d{1,2}$`)
+
+// Server is the upgrade-agent server.
+type Server struct {
+	file       file.Handler
+	grpcServer serveStopper
+	log        *logger.Logger
+	upgradeproto.UnimplementedUpdateServer
+}
+
+// New creates a new upgrade-agent server.
+func New(log *logger.Logger, fileHandler file.Handler) (*Server, error) {
+	log = log.Named("upgradeServer")
+
+	server := &Server{
+		log:  log,
+		file: fileHandler,
+	}
+
+	grpcServer := grpc.NewServer(
+		log.Named("gRPC").GetServerUnaryInterceptor(),
+	)
+	upgradeproto.RegisterUpdateServer(grpcServer, server)
+
+	server.grpcServer = grpcServer
+	return server, nil
+}
+
+// Run starts the upgrade-agent server on the given port, using the provided protocol and socket address.
+func (s *Server) Run(protocol string, sockAddr string) error {
+	grpcServer := grpc.NewServer()
+
+	upgradeproto.RegisterUpdateServer(grpcServer, s)
+
+	cleanup := func() error {
+		if _, err := os.Stat(sockAddr); err == nil {
+			if err := os.RemoveAll(sockAddr); err != nil {
+				return err
+			}
+		}
```

Comment:
```suggestion
	err := os.RemoveAll(sockAddr)
	if errors.Is(err, fs.ErrNotExist) {
		return nil
	} else if err != nil {
		return err
	}
```
Using `os.Stat` when followed by anything like `os.Open`, `os.Remove`, or similar is an antipattern since it leads to [TOCTOU](https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use) issues. Those file methods are small wrappers around their syscall counterparts like [open](https://man7.org/linux/man-pages/man2/open.2.html), [close](https://man7.org/linux/man-pages/man2/close.2.html), and so on. Those syscalls have defined error cases with which one can avoid a prior stat usage. Those errors are also exposed as Go error types. I.e. in 90% of the cases a `os.Stat` can be replaced with a `errors.Is(err, fs.ErrNotExist)` :smile: